torch/csrc/jit/irparser.cpp

```diff
@@ -0,0 +1,432 @@
+#include <torch/csrc/jit/irparser.h>
+#include <torch/csrc/jit/ir.h>
+#include <torch/csrc/jit/script/lexer.h>
+#include <torch/csrc/jit/script/parse_string_literal.h>
+
+#include <string>
+#include <vector>
+
+namespace torch {
+namespace jit {
+namespace script {
+
+struct ParsedValue;
+struct ParsedLiteral;
+
+class IRParser {
+  friend void parseIR(const std::string& str, torch::jit::Graph* graph);
+  IRParser(const std::string& str, torch::jit::Graph* graph)
+      : L(str), g(graph) {}
+
+  ParsedValue parseParameter();
+  ParsedLiteral parseScalarLiteral(Node* n);
+
+  void parse();
+  void parseGraphInputs();
+  void parseReturnStmt();
+
+  void parseBlocks(Node* parentNode);
+  void parseBlock(Node* parentNode);
+  void parseBlockInputs(Block* b);
+  void parseBlockOutputs(Block* b);
+
+  void parseStmtsList(Block* b);
+  void parseStmt(Block* b);
+  void parseStmtOutputs(std::vector<ParsedValue>* outs);
+  std::string parseStmtName();
+  void parseStmtInputs(Node* n);
+  void parseAttrs(Node* n);
+  void parseAttr(Node* n);
+
+  void parseList(
+      int begin,
+      int sep,
+      int end,
+      const std::function<void()>& callback);
+
+  torch::jit::script::Lexer L;
+  torch::jit::Graph* g = nullptr;
+  std::unordered_map<std::string, Value*> vmap;
+};
+
+struct ParsedLiteral {
+  AttributeKind k = AttributeKind::t;
+
+  long i;
+  std::string s;
+  double f;
+  std::vector<long> is;
+  std::vector<std::string> ss;
+  std::vector<double> fs;
+};
+
+struct ParsedValue {
+  ParsedValue(){};
+  std::string name;
+  std::string type;
+};
+
+void parseIR(const std::string& str, torch::jit::Graph* graph) {
+  torch::jit::script::IRParser p(str, graph);
+  p.parse();
+}
+
+TypePtr parseType(const std::string& s) {
+  if (s == "Tensor" || s == "NoType") {
```

Comment:
Thanks for the suggestion, I'll take a look at those parser and see what can be shared.

Reply:
I don't think it's possible currently, since we have types internally that are not user-exposed - CompleteTensorType, for example.

Maybe in the future if we expose annotating detailed Tensor Types than it would work. But as of now I'm not sure that much can be shared.

Reply:
We should not end up with a fourth type parser. I think it would be best to share the schema parser's ```std::pair<TypePtr, c10::optional<AliasInfo>> parseType()```, which can be refactored to be a common component. It is the closest in syntax to what the IR printer prints, and we can add cases for DimensionedTensor and CompleteTensor that are feature-gated by a boolean. For now this code is fine, as long as the followup that merges it happens.